checker/src/org/checkerframework/checker/i18nformatter/I18nFormatterTreeUtil.java

```diff
@@ -254,49 +248,46 @@ public String toString() {
-                        // Invalid FormatFor invocation
-                        return;
-                    }
-                    FlowExpressionContext flowExprContext =
-                            FlowExpressionContext.buildContextForMethodUse(
-                                    node, checker.getContext());
-                    String formatforArg =
-                            AnnotationUtils.getElementValue(
-                                    paramType.getAnnotation(I18nFormatFor.class),
-                                    "value",
-                                    String.class,
-                                    false);
-                    if (flowExprContext != null) {
-                        try {
-                            paramArg =
-                                    FlowExpressionParseUtil.parse(
-                                            formatforArg,
-                                            flowExprContext,
-                                            atypeFactory.getPath(tree),
-                                            true);
-                            paramIndex = flowExprContext.arguments.indexOf(paramArg);
-                        } catch (FlowExpressionParseException e) {
-                            // report errors here
-                            checker.report(
-                                    org.checkerframework.framework.source.Result.failure(
-                                            "i18nformat.invalid.formatfor"),
-                                    tree);
-                        }
-                    }
-                    break;
+                if (!typeMirrorToClass(paramType.getUnderlyingType()).equals(String.class)) {
+                    // Invalid FormatFor invocation
+                    return;
+                }
+                FlowExpressionContext flowExprContext =
+                        FlowExpressionContext.buildContextForMethodUse(tree, checker.getContext());
+                String formatforArg =
+                        AnnotationUtils.getElementValue(
+                                paramType.getAnnotation(I18nFormatFor.class),
+                                "value",
+                                String.class,
+                                false);
+                // Parse the expression and issue errors.
+                try {
+                    FlowExpressionParseUtil.parse(
+                            formatforArg, flowExprContext, atypeFactory.getPath(tree), true);
+                } catch (FlowExpressionParseException e) {
+                    // report errors here
+                    checker.report(
+                            org.checkerframework.framework.source.Result.failure(
+                                    "i18nformat.invalid.formatfor"),
+                            tree);
+                    return;
+                }
+                if (formatforArg.startsWith("#")) {
```

Comment:
In https://github.com/typetools/checker-framework/pull/1844#discussion_r170652526 you said `The current code allows any FlowExpression. Your example code does not. I don't know the code well enough to know if restricting to just parameter references is ok.`
This check enforces that the string is of the format `#integer`, so it equally restricts what flow expressions are accepted.
So why not replace lines 275 - 289 by

```java
if (formatforArg.startsWith("#")) {
    // formatArg should be #Number, with 1-indexing.
    try {
        paramIndex = Integer.valueOf(formatforArg.substring(1)) - 1;
    } catch (NumberFormatException nfe) {
        checker.report(Invalid format);
    }
    if (paramIndex < 0 || paramIndex > method.getParameters().size()) {
        checker.report(Invalid index);
    }
}
```

Alternatively, could the old code be used and the ambiguity between 1 and #1 be resolved by checking for #, without re-parsing the integer.

Reply:
I don't know if other code relies on this code to issue parse error of any flow expression.  I would prefer to keep the code as close to the old code as possible.

>Alternatively, could the old code be used and the ambiguity between 1 and #1 be resolved by checking for #, without re-parsing the integer.

The problem is not the difference between `1` and `#1`, it's between to arguments that parse to equal Receivers.